### PR TITLE
8270160: Remove redundant bounds check from AbstractStringBuilder.charAt()

### PR DESCRIPTION
**Pull request by stsypanov**
`AbstractStringBuilder.charAt(int)` does bounds check before calling `charAt()` (for non-latin Strings):
```java
@Override
public char charAt(int index) {
    checkIndex(index, count);
    if (isLatin1()) {
        return (char)(value[index] & 0xff);
    }
    return StringUTF16.charAt(value, index);
}
```
This can be improved by removing bounds check from ASB.charAt() in favour of one in String*.charAt(). This gives slight improvement:
```
before
Benchmark                           Mode  Cnt  Score   Error  Units
StringBuilderCharAtBenchmark.latin  avgt   50  2,827 ± 0,024  ns/op
StringBuilderCharAtBenchmark.utf    avgt   50  2,985 ± 0,020  ns/op

after
Benchmark                           Mode  Cnt  Score   Error  Units
StringBuilderCharAtBenchmark.latin  avgt   50  2,434 ± 0,004  ns/op
StringBuilderCharAtBenchmark.utf    avgt   50  2,631 ± 0,004  ns/op
```

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8270160](https://bugs.openjdk.java.net/browse/JDK-8270160): Remove redundant bounds check from AbstractStringBuilder.charAt()

### Reviewers
 * [Claes Redestad](https://openjdk.java.net/census#redestad) (@cl4es - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4738/head:pull/4738` \
`$ git checkout pull/4738`

Update a local copy of the PR: \
`$ git checkout pull/4738` \
`$ git pull https://git.openjdk.java.net/jdk pull/4738/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4738`

View PR using the GUI difftool: \
`$ git pr show -t 4738`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4738.diff">https://git.openjdk.java.net/jdk/pull/4738.diff</a>

</details>

### CODE REVIEW

File: src/java.base/share/classes/java/lang/AbstractStringBuilder.java [hunk 1]
```diff
@@ -224,7 +224,7 @@ public void ensureCapacity(int minimumCapacity) {
      */
     private void ensureCapacityInternal(int minimumCapacity) {
         // overflow-conscious code
-        int oldCapacity = value.length >> coder;
+        int oldCapacity = capacity();
```
Comment (cl4es) [inline src/java.base/share/classes/java/lang/AbstractStringBuilder.java:227]: This doesn't seem related to the intent of the patch, and might distort startup profiles a bit.

File: src/java.base/share/classes/java/lang/StringLatin1.java [hunk 2]
```diff
@@ -46,7 +44,7 @@
 
     public static char charAt(byte[] value, int index) {
         checkIndex(index, value.length);
-        return (char)(value[index] & 0xff);
+        return getChar(value, index);
```
Comment (cl4es) [inline src/java.base/share/classes/java/lang/StringLatin1.java:47]: Seems unrelated to the gist of the patch, and again might skew startup profiles.

**Review by cl4es — CHANGES_REQUESTED** [review]

File: src/java.base/share/classes/java/lang/AbstractStringBuilder.java (hunk 1 above)
Comment (stsypanov) [reply src/java.base/share/classes/java/lang/AbstractStringBuilder.java]: Reverted

File: src/java.base/share/classes/java/lang/StringLatin1.java (hunk 2 above)
Comment (stsypanov) [reply src/java.base/share/classes/java/lang/StringLatin1.java]: Reverted

**Review by cl4es — APPROVED** [review]
